### PR TITLE
🏗️: walled garden of private API

### LINKED ISSUE #9: Use walled-garden approach
The following approach allows for very exact control over the public API while also making devs' lives easy with a full private API and non-prefix-underscored file names.

```
src/
    quantity/
        __init__.py  # defines the public API
        api.py  # e.g. a public module with public API that contains the API protocols.
        _src/  # private API walled garden. Everything here is private and easily worked on by devs w/out public API consequences.
            base.py  # named whatever we want. no pre-fix underscores.
            core.py  # named whatever we want. no pre-fix underscores.
            api.py  # named whatever we want. no pre-fix underscores.
            errors.py  # named whatever we want. no pre-fix underscores.
```

`__init__.py` and the example `api.py` are very simple modules that define the public API. For example

```
# __init__.py
"""User-facing docstring for quantity package"""

__all__ = ["AbstractQuantity", "Quantity", ...]

from ._src.base import AbstractQuantity
from ._src.core import Quantity
```

```
# api.py
"""User-facing docstring for quantity.api module"""

__all__ = ["Quantity", "ArrayQuantity"]

from ._src.api import Quantity, ArrayQuantity
```

Check out https://github.com/GalacticDynamics/unxt/tree/main/src/unxt, where I've implemented this code pattern. It's worked **wonderfully**. The API is 100% clear and 100% complies with PEP 8 standards — a great benefit individually and synergistically for users, developers, and code tools. With just one underscore-prefixed module this pattern completely separates public from private API and makes life very easy for developers who never need to worry about moving classes, renaming / deleting files. Private files  also aren't themselves underscore-prefixed, which is a major bonus. Another bonus is that the `__all__` in the private modules can define a private API, e.g. functions that are marked for use in other private modules, but not in the public API, which has its own clear definition.

### PR DESCRIPTION
**Pull request by nstarman**
Also moving a utility function, without making it public.

This PR shows how we can separate the public from private API, enabling easy dev on the private API without impacting the public API.

Closes #9

### CODE REVIEW

**Review by mhvk — COMMENTED** [review]
As said in the issue, like this approach in principle. Is there precedent for `_src` rather than, say, `_impl` or `_code` or something? Obviously, another benefit of this being private is that it doesn't matter what the name is, and `_src` is nice and short....

**Comment by nstarman** [comment]
I don't think there's a strong precedent.

File: src/quantity/_src/__init__.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+"""Quantity 2.0.
+
+This `_src` module implements the private API.
+The public API is available in the top-level `quantity` module.
+
+"""
+# Copyright (c) 2024 Astropy Developers. All rights reserved.
+
+from __future__ import annotations
```
Comment (mhvk) [inline src/quantity/_src/__init__.py:9]: Leftover?

**Review by mhvk — APPROVED** [review]
OK, let's go with `_src` - definitely better than `_lib` (even if I might have preferred `_impl` or `_implementation` myself).

One in-line comment, which may be worth addressing (but is not critical in any way).

File: src/quantity/_src/__init__.py (hunk 1 above)
Comment (nstarman) [reply src/quantity/_src/__init__.py]: Removing.

**Pull request merged**